File: hdf-utilities/test/status.test.ts
import { describe, it, expect } from 'vitest';
import {
  STATUS_SEVERITY_ORDER,
  statusRank,
  worstStatus,
  governingStatusOverride,
  governingOverrideIndex,
  computeEffectiveStatus,
} from '../src/status/index.js';

const REF = '2026-01-01T00:00:00Z';
const FAR_FUTURE = '2099-12-31T00:00:00Z';
const LONG_AGO = '2020-01-01T00:00:00Z';
const APPLIED_OLD = '2025-01-01T00:00:00Z';
const APPLIED_NEW = '2025-06-01T00:00:00Z';

describe('statusRank / STATUS_SEVERITY_ORDER', () => {
  it('ranks the canonical ordering, higher = worse', () => {
    expect(STATUS_SEVERITY_ORDER).toEqual([
      'error',
      'failed',
      'passed',
      'notApplicable',
      'notReviewed',
    ]);
    expect(statusRank('error')).toBe(4);
    expect(statusRank('failed')).toBe(3);
    expect(statusRank('passed')).toBe(2);
    expect(statusRank('notApplicable')).toBe(1);
    expect(statusRank('notReviewed')).toBe(0);
    expect(statusRank('bogus')).toBe(-1);
  });
});

describe('worstStatus', () => {
  it('rolls up worst-wins per the canonical ordering', () => {
    expect(worstStatus([])).toBe('notReviewed');
    expect(worstStatus(['passed'])).toBe('passed');
    expect(worstStatus(['passed', 'failed'])).toBe('failed');
    expect(worstStatus(['failed', 'error', 'passed'])).toBe('error');
    expect(worstStatus(['notReviewed', 'passed'])).toBe('passed');
    expect(worstStatus(['notReviewed', 'notApplicable'])).toBe('notApplicable');
    expect(worstStatus(['bogus', 'passed'])).toBe('passed');
    expect(worstStatus(['bogus'])).toBe('notReviewed');
  });
});

describe('governingOverrideIndex', () => {
  const all = () => true;

  it('selects the most recently applied eligible non-expired override', () => {
    const overrides = [
      { appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE },
      { appliedAt: APPLIED_NEW, expiresAt: FAR_FUTURE },
    ];
    expect(governingOverrideIndex(overrides, all, REF)).toBe(1);
  });

  it('skips ineligible overrides', () => {
    const overrides = [
      { appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE },
      { appliedAt: APPLIED_NEW, expiresAt: FAR_FUTURE },
    ];
    expect(governingOverrideIndex(overrides, (i) => i === 0, REF)).toBe(0);
  });

  it('skips expired overrides', () => {
    const overrides = [
      { appliedAt: APPLIED_NEW, expiresAt: LONG_AGO },
      { appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE },
    ];
    expect(governingOverrideIndex(overrides, all, REF)).toBe(1);
  });

  it('treats a missing expiresAt as never expiring', () => {
    expect(governingOverrideIndex([{ appliedAt: APPLIED_OLD }], all, REF)).toBe(0);
  });

  it('returns -1 when none are eligible', () => {
    const overrides = [{ appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE }];
    expect(governingOverrideIndex(overrides, () => false, REF)).toBe(-1);
  });
});

describe('governingStatusOverride', () => {
  it('selects the most recently applied non-expired override', () => {
    const governing = governingStatusOverride(
      [
        { status: 'notApplicable', appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE },
        { status: 'failed', appliedAt: APPLIED_NEW, expiresAt: FAR_FUTURE },
      ],
      REF
    );
    expect(governing?.status).toBe('failed');
  });

  it('skips expired overrides', () => {
    const governing = governingStatusOverride(
      [
        { status: 'failed', appliedAt: APPLIED_NEW, expiresAt: LONG_AGO },
        { status: 'notApplicable', appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE },
      ],
      REF
    );
    expect(governing?.status).toBe('notApplicable');
  });

  it('treats a missing expiresAt as never expiring', () => {
    const governing = governingStatusOverride([{ status: 'passed', appliedAt: APPLIED_OLD }], REF);
    expect(governing?.status).toBe('passed');
  });

  it('returns undefined when all overrides are expired or statusless', () => {
    expect(
      governingStatusOverride([{ status: 'passed', appliedAt: APPLIED_OLD, expiresAt: LONG_AGO }], REF)
    ).toBeUndefined();
    expect(
      governingStatusOverride([{ appliedAt: APPLIED_NEW, expiresAt: FAR_FUTURE }], REF)
    ).toBeUndefined();
  });
});

describe('computeEffectiveStatus', () => {
  it('forces notApplicable at impact zero regardless of results', () => {
    expect(computeEffectiveStatus({ impact: 0, resultStatuses: ['failed'] }, REF)).toBe(
      'notApplicable'
    );
  });

  it('lets the governing override win over results and effectiveStatus', () => {
    expect(
      computeEffectiveStatus(
        {
          impact: 0.7,
          effectiveStatus: 'failed',
          resultStatuses: ['failed'],
          overrides: [{ status: 'notApplicable', appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE }],
        },
        REF
      )
    ).toBe('notApplicable');
  });

  it('honors effectiveStatus only when no overrides are present', () => {
    expect(
      computeEffectiveStatus(
        { impact: 0.7, effectiveStatus: 'passed', resultStatuses: ['failed'] },
        REF
      )
    ).toBe('passed');
  });

  it('recomputes from results when every override has expired', () => {
    expect(
      computeEffectiveStatus(
        {
          impact: 0.7,
          effectiveStatus: 'passed',
          resultStatuses: ['failed'],
          overrides: [{ status: 'notApplicable', appliedAt: APPLIED_OLD, expiresAt: LONG_AGO }],
        },
        REF
      )
    ).toBe('failed');
  });

  it('rolls results up worst-wins and defaults empty results to notReviewed', () => {
    expect(
      computeEffectiveStatus({ impact: 0.5, resultStatuses: ['passed', 'error', 'failed'] }, REF)
    ).toBe('error');
    expect(computeEffectiveStatus({ impact: 0.5 }, REF)).toBe('notReviewed');
  });

  it('falls back to the wall clock when no reference timestamp is given', () => {
    expect(
      computeEffectiveStatus({
        impact: 0.7,
        resultStatuses: ['failed'],
        overrides: [{ status: 'notApplicable', appliedAt: APPLIED_OLD, expiresAt: FAR_FUTURE }],
      })
    ).toBe('notApplicable');
  });
});

File: hdf-utilities/go/status_test.go
package hdfutil

import (
	"testing"
	"time"
)

var (
	statusRef  = time.Date(2026, 1, 1, 0, 0, 0, 0, time.UTC)
	farFuture  = time.Date(2099, 12, 31, 0, 0, 0, 0, time.UTC)
	longAgo    = time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)
	appliedOld = time.Date(2025, 1, 1, 0, 0, 0, 0, time.UTC)
	appliedNew = time.Date(2025, 6, 1, 0, 0, 0, 0, time.UTC)
)

func TestStatusRank(t *testing.T) {
	// Canonical ordering per status-determination.md: higher rank = worse.
	cases := []struct {
		status string
		rank   int
	}{
		{"error", 4},
		{"failed", 3},
		{"passed", 2},
		{"notApplicable", 1},
		{"notReviewed", 0},
	}
	for _, c := range cases {
		if got := StatusRank(c.status); got != c.rank {
			t.Errorf("StatusRank(%q) = %d, want %d", c.status, got, c.rank)
		}
	}
	if got := StatusRank("bogus"); got != -1 {
		t.Errorf("StatusRank(bogus) = %d, want -1", got)
	}
}

func TestWorstStatus(t *testing.T) {
	cases := []struct {
		name     string
		statuses []string
		want     string
	}{
		{"empty", nil, "notReviewed"},
		{"single passed", []string{"passed"}, "passed"},
		{"failed beats passed", []string{"passed", "failed"}, "failed"},
		{"error beats failed", []string{"failed", "error", "passed"}, "error"},
		{"passed beats notReviewed", []string{"notReviewed", "passed"}, "passed"},
		{"notApplicable beats notReviewed", []string{"notReviewed", "notApplicable"}, "notApplicable"},
		{"all notReviewed", []string{"notReviewed", "notReviewed"}, "notReviewed"},
		{"unknown statuses ignored", []string{"bogus", "passed"}, "passed"},
		{"only unknown statuses", []string{"bogus"}, "notReviewed"},
	}
	for _, c := range cases {
		if got := WorstStatus(c.statuses); got != c.want {
			t.Errorf("%s: WorstStatus(%v) = %q, want %q", c.name, c.statuses, got, c.want)
		}
	}
}

func TestGoverningStatusOverride(t *testing.T) {
	waived := "notApplicable"
	failedStr := "failed"

	t.Run("most recent AppliedAt wins among non-expired", func(t *testing.T) {
		overrides := []StatusOverrideInput{
			{Status: waived, AppliedAt: appliedOld, ExpiresAt: farFuture},
			{Status: failedStr, AppliedAt: appliedNew, ExpiresAt: farFuture},
		}
		got := GoverningStatusOverride(overrides, statusRef)
		if got == nil || got.Status != failedStr {
			t.Fatalf("got %+v, want the later-applied override", got)
		}
	})

	t.Run("expired overrides are skipped", func(t *testing.T) {
		overrides := []StatusOverrideInput{
			{Status: failedStr, AppliedAt: appliedNew, ExpiresAt: longAgo},
			{Status: waived, AppliedAt: appliedOld, ExpiresAt: farFuture},
		}
		got := GoverningStatusOverride(overrides, statusRef)
		if got == nil || got.Status != waived {
			t.Fatalf("got %+v, want the non-expired override", got)
		}
	})

	t.Run("zero ExpiresAt never expires", func(t *testing.T) {
		overrides := []StatusOverrideInput{{Status: waived, AppliedAt: appliedOld}}
		if got := GoverningStatusOverride(overrides, statusRef); got == nil {
			t.Fatal("zero ExpiresAt must count as non-expired")
		}
	})

	t.Run("all expired yields nil", func(t *testing.T) {
		overrides := []StatusOverrideInput{{Status: waived, AppliedAt: appliedOld, ExpiresAt: longAgo}}
		if got := GoverningStatusOverride(overrides, statusRef); got != nil {
			t.Fatalf("got %+v, want nil", got)
		}
	})

	t.Run("statusless overrides are not governing", func(t *testing.T) {
		overrides := []StatusOverrideInput{{AppliedAt: appliedNew, ExpiresAt: farFuture}}
		if got := GoverningStatusOverride(overrides, statusRef); got != nil {
			t.Fatalf("got %+v, want nil for override without status", got)
		}
	})
}

func TestGoverningOverrideIndex(t *testing.T) {
	all := func(int) bool { return true }

	t.Run("most recent AppliedAt wins among eligible non-expired", func(t *testing.T) {
		overrides := []StatusOverrideInput{
			{AppliedAt: appliedOld, ExpiresAt: farFuture},
			{AppliedAt: appliedNew, ExpiresAt: farFuture},
		}
		if got := GoverningOverrideIndex(overrides, all, statusRef); got != 1 {
			t.Fatalf("got %d, want 1 (later-applied)", got)
		}
	})

	t.Run("ineligible overrides are skipped", func(t *testing.T) {
		overrides := []StatusOverrideInput{
			{AppliedAt: appliedOld, ExpiresAt: farFuture},
			{AppliedAt: appliedNew, ExpiresAt: farFuture},
		}
		onlyFirst := func(i int) bool { return i == 0 }
		if got := GoverningOverrideIndex(overrides, onlyFirst, statusRef); got != 0 {
			t.Fatalf("got %d, want 0 (only eligible)", got)
		}
	})

	t.Run("expired overrides are skipped", func(t *testing.T) {
		overrides := []StatusOverrideInput{
			{AppliedAt: appliedNew, ExpiresAt: longAgo},
			{AppliedAt: appliedOld, ExpiresAt: farFuture},
		}
		if got := GoverningOverrideIndex(overrides, all, statusRef); got != 1 {
			t.Fatalf("got %d, want 1 (non-expired)", got)
		}
	})

	t.Run("zero ExpiresAt never expires", func(t *testing.T) {
		overrides := []StatusOverrideInput{{AppliedAt: appliedOld}}
		if got := GoverningOverrideIndex(overrides, all, statusRef); got != 0 {
			t.Fatalf("got %d, want 0", got)
		}
	})

	t.Run("none eligible yields -1", func(t *testing.T) {
		overrides := []StatusOverrideInput{{AppliedAt: appliedOld, ExpiresAt: farFuture}}
		none := func(int) bool { return false }
		if got := GoverningOverrideIndex(overrides, none, statusRef); got != -1 {
			t.Fatalf("got %d, want -1", got)
		}
	})
}

func TestComputeEffectiveStatus(t *testing.T) {
	waived := "notApplicable"

	t.Run("impact zero always notApplicable", func(t *testing.T) {
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:         0,
			ResultStatuses: []string{"failed"},
		}, statusRef)
		if got != "notApplicable" {
			t.Errorf("got %q, want notApplicable", got)
		}
	})

	t.Run("governing override wins over results and effectiveStatus", func(t *testing.T) {
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:          0.7,
			EffectiveStatus: "failed",
			ResultStatuses:  []string{"failed"},
			Overrides: []StatusOverrideInput{
				{Status: waived, AppliedAt: appliedOld, ExpiresAt: farFuture},
			},
		}, statusRef)
		if got != waived {
			t.Errorf("got %q, want %q", got, waived)
		}
	})

	t.Run("effectiveStatus honored only when no overrides present", func(t *testing.T) {
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:          0.7,
			EffectiveStatus: "passed",
			ResultStatuses:  []string{"failed"},
		}, statusRef)
		if got != "passed" {
			t.Errorf("got %q, want passed", got)
		}
	})

	t.Run("expired overrides invalidate stale effectiveStatus", func(t *testing.T) {
		// effectiveStatus is derived state; when its overrides have all
		// expired it goes stale, so the result rollup wins.
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:          0.7,
			EffectiveStatus: "passed",
			ResultStatuses:  []string{"failed"},
			Overrides: []StatusOverrideInput{
				{Status: waived, AppliedAt: appliedOld, ExpiresAt: longAgo},
			},
		}, statusRef)
		if got != "failed" {
			t.Errorf("got %q, want failed (stale effectiveStatus recomputed)", got)
		}
	})

	t.Run("rollup worst-wins", func(t *testing.T) {
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:         0.5,
			ResultStatuses: []string{"passed", "error", "failed"},
		}, statusRef)
		if got != "error" {
			t.Errorf("got %q, want error", got)
		}
	})

	t.Run("no results notReviewed", func(t *testing.T) {
		got := ComputeEffectiveStatus(EffectiveStatusInput{Impact: 0.5}, statusRef)
		if got != "notReviewed" {
			t.Errorf("got %q, want notReviewed", got)
		}
	})

	t.Run("zero reference time uses wall clock", func(t *testing.T) {
		// A far-future expiry must be non-expired against the wall clock too.
		got := ComputeEffectiveStatus(EffectiveStatusInput{
			Impact:         0.7,
			ResultStatuses: []string{"failed"},
			Overrides: []StatusOverrideInput{
				{Status: waived, AppliedAt: appliedOld, ExpiresAt: farFuture},
			},
		}, time.Time{})
		if got != waived {
			t.Errorf("got %q, want %q", got, waived)
		}
	})
}
